eclass/kernel-build.eclass

```diff
@@ -534,40 +538,85 @@ kernel-build_src_install() {
 			dracut "${dracut_args[@]}" "${image%/*}/initrd" ||
 				die "Failed to generate initramfs"
 
-			local ukify_args=(
+			local ukify_base_args=(
 				--linux="${image}"
 				--initrd="${image%/*}/initrd"
-				--cmdline="${KERNEL_GENERIC_UKI_CMDLINE}"
 				--uname="${KV_FULL}"
-				--output="${image%/*}/uki.efi"
 			)
 
-			if [[ ${KERNEL_IUSE_MODULES_SIGN} ]] && use secureboot; then
-				ukify_args+=(
-					--signtool=sbsign
-					--secureboot-private-key="${SECUREBOOT_SIGN_KEY}"
-					--secureboot-certificate="${SECUREBOOT_SIGN_CERT}"
+			local pcrpkey=${ED}${kernel_dir}/certs/signing_key.x509
+			if [[ -r ${pcrpkey} ]]; then
+				ukify_base_args+=(
+					--pcrpkey="${pcrpkey}"
 				)
-				if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
-					ukify_args+=(
-						--signing-engine="pkcs11"
-					)
-				else
-					# Sytemd-measure does not currently support pkcs11
-					ukify_args+=(
+			fi
+
+			ukify build "${ukify_base_args[@]}" \
+				--output="${T}/base-uki.efi" ||
+					die "Failed to generate base UKI"
+
+
+			local ukify_extend_args=()
+			if [[ ${KERNEL_IUSE_MODULES_SIGN} ]] && use secureboot; then
+				ukify_extend_args+=(
 						--measure
-						--pcrpkey="${ED}${kernel_dir}/certs/signing_key.x509"
 						--pcr-private-key="${SECUREBOOT_SIGN_KEY}"
 						--phases="enter-initrd"
 						--pcr-private-key="${SECUREBOOT_SIGN_KEY}"
 						--phases="enter-initrd:leave-initrd enter-initrd:leave-initrd:sysinit enter-initrd:leave-initrd:sysinit:ready"
+				)
+				if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+					ukify_extend_args+=(
+						--signing-engine="pkcs11"
+						--pcr-public-key="${SECUREBOOT_SIGN_CERT}"
 					)
 				fi
 			fi
 
-			# systemd<255 does not install ukify in /usr/bin
-			PATH="${PATH}:${BROOT}/usr/lib/systemd:${BROOT}/lib/systemd" \
-				ukify build "${ukify_args[@]}" || die "Failed to generate UKI"
+			# Note, we cannot use an associative array here because those are
+			# not ordered.
+			local profiles=()
+			local cmdlines=()
+
+			# If defined, make the custom entry the first and default
+			if [[ -n ${KERNEL_GENERIC_UKI_CMDLINE} ]]; then
+				profiles+=(
+					"TITLE=Custom\nID=custom"
```

Comment:
> "Custom" sounds a bit weird, but I don't have a better idea. @thesamesam ?

Something else I considered was "user specified", but this I found misleading because it seemed to imply that the user could then specify it from the bootloader which is not really what this does. Then I considered something like "build time specified" but this also is strange since all options are build time specified. In the end I could not really find something that properly describes this in one or two words, so I stayed with "custom", but this is also vague indeed.

Reply:
We could also consider simply dropping the custom `KERNEL_GENERIC_UKI_CMDLINE` as a user option. The main reason I added it in the first place was as a workaround for [Bug 921195](https://bugs.gentoo.org/921195), i.e. as a workaround for that it was not possible to boot with both secure boot and module signature verification. This we can now resolve properly due to the introduction of multi-profile UKIs. So the original reason for adding this workaround is now gone. 

That being said, I don't know if people are using  `KERNEL_GENERIC_UKI_CMDLINE` for other things that I have not considered. And as a concept, the possibility for users to define some custom profile for the UKI is kinda nice.

Reply:
What about "User built-in"?

Reply:
"Specified at compile-time"?